src/main/host/syscall/signal.c

```diff
@@ -16,6 +16,34 @@
 #include "main/host/thread.h"
 #include "main/utility/syscall.h"
 
+// This should be compatible with the kernel's sigaction struct, which is what
+// the syscalls take in cases where the userspace wrappers take `struct
+// sigaction`.
+//
+// When reading/writing a userspace pointer, use kernel_sigaction_size to
+// calculate the true size. This is needed because the last field, `sa_mask`,
+// is actually variable size, and is specified with another parameter in the syscall.
+struct kernel_sigaction {
+    void* handler;
+    unsigned long sa_flags;
+    void (*sa_restorer)(void);
+    sigset_t sa_mask;
+};
```

Comment:
Is there a reason to use this struct rather than including the one from ` linux/signal.h`?

Reply:
Short answer: it's not there :). I expanded on the comment a bit.

```
$ grep -R kernel_sigaction /usr/include
<no output>
```

Reply:
It seems to be there for me on ubuntu 20.04:

```c
#include <stddef.h>
#include <linux/signal.h>

int main() {
    struct sigaction test;
    return 0;
}
```

And I think it's just named `sigaction` in the kernel?

Reply:
The glibc public interface uses `struct sigaction`; the kernel syscall API uses `struct kernel_sigaction`

Reply:
I thought they both used the name `struct sigaction`, but glibc [uses the name `struct kernel_sigaction` internally](https://code.woboq.org/userspace/glibc/sysdeps/unix/sysv/linux/kernel_sigaction.h.html) to refer to their copy. But either way I think it's fine to redefine it, it shouldn't change anytime soon.

Reply:
Sadly they are different structs :). `kernel_sigaction` has the mask last (presumably to support the run-time mask-size without changing the offsets of the other members).

https://github.com/sporksmith/glibc/blob/master/sysdeps/unix/sysv/linux/bits/sigaction.h#L27
```
struct sigaction
  {
    /* Signal handler.  */
#if defined __USE_POSIX199309 || defined __USE_XOPEN_EXTENDED
    union
      {
	/* Used if SA_SIGINFO is not set.  */
	__sighandler_t sa_handler;
	/* Used if SA_SIGINFO is set.  */
	void (*sa_sigaction) (int, siginfo_t *, void *);
      }
    __sigaction_handler;
# define sa_handler	__sigaction_handler.sa_handler
# define sa_sigaction	__sigaction_handler.sa_sigaction
#else
    __sighandler_t sa_handler;
#endif

    /* Additional set of signals to be blocked.  */
    __sigset_t sa_mask;

    /* Special flags.  */
    int sa_flags;

    /* Restore handler.  */
    void (*sa_restorer) (void);
  };
```

https://github.com/sporksmith/glibc/blob/master/sysdeps/unix/sysv/linux/kernel_sigaction.h#L9
```
struct kernel_sigaction
{
  __sighandler_t k_sa_handler;
  unsigned long sa_flags;
#ifdef HAS_SA_RESTORER
  void (*sa_restorer) (void);
#endif
  /* glibc sigset is larger than kernel expected one, however sigaction
     passes the kernel expected size on rt_sigaction syscall.  */
  sigset_t sa_mask;
};
```

Reply:
Actually, looking at it again, I think the `sigaction` syscall does take `struct sigaction`, and it's only `sigaction_rt` that takes `struct sigaction`. i.e. they used to be the same and then rearranged it for future expandability (while leaving the glibc interface the same)

Reply:
Will verify and prep a follow-up. I think modern libc always uses the `rt_sigaction` syscall, but older ones might still use `sigaction`...

Reply:
Yup.

https://github.com/sporksmith/linux/blob/144c79ef33536b4ecb4951e07dbc1f2b7fa99d32/kernel/signal.c#L4360
```
SYSCALL_DEFINE3(sigaction, int, sig,
		const struct old_sigaction __user *, act,
	        struct old_sigaction __user *, oact)
{
```

https://github.com/sporksmith/linux/blob/master/include/linux/compat.h#L369
```
struct compat_old_sigaction {
	compat_uptr_t			sa_handler;
	compat_old_sigset_t		sa_mask;
	compat_ulong_t			sa_flags;
	compat_uptr_t			sa_restorer;
};
```